src/traceme.c

```diff
@@ -273,49 +313,104 @@ tracer_thread(void *arg)
     attr.precise_ip = 3;
 
     // Acquire file descriptor through which to talk to Intel PT.
-    struct tracer_ctx *tr_ctx = (struct tracer_ctx *) arg;
-    tr_ctx->perf_fd = syscall(SYS_perf_event_open, &attr,
-        tr_ctx->target_pid, -1, -1, 0);
-    if (tr_ctx->perf_fd == -1) {
-        err(EXIT_FAILURE, "syscall");
+    ret = syscall(SYS_perf_event_open, &attr, target_pid, -1, -1, 0);
+
+clean:
+    if (fclose(pt_type_file) == -1) {
+        ret = -1;
     }
 
+    return ret;
+}
+
+/*
+ * Set up Intel PT buffers and start a poll() loop for reading out the trace.
+ */
+static void *
+tracer_thread(void *arg)
+{
+    DEBUG("tracer init");
+
+    /* Copy args in */
+    struct tracer_thread_args thr_args;
+    memcpy(&thr_args, (struct tracer_thread_args *) arg, sizeof(thr_args));
```

Comment:
We are copying a pointer to a string on the rust stack here. That pointer becomes invalid after `start_tracing` in rust has returned. Although use of the pointer is synchronised, and we don't use the pointer after, I wonder if we should try and eliminate that pointer hanging around...

Reply:
The pointer is the output filename, btw. We could open the file in the caller, and pass a fd instead. I'd be inclined to have the parent close the fd too if that were our plan.